File: scripts/run-tailoring-eval.py
import json
from pathlib import Path
import requests
from requests.auth import HTTPBasicAuth

BASE_URL = "http://127.0.0.1:3005"
AUTH = HTTPBasicAuth("jobops", "ajhg0Pdm1qAuVWYtleJCN67CKnaOQdta")
FIXTURE_PATH = Path("/opt/job-ops/scripts/tailoring-eval-fixtures/denmark-planning.json")


def contains_keyword(text: str, keyword: str) -> bool:
    return keyword.lower() in text.lower()


def flatten_payload(data: dict) -> str:
    skills = []
    for group in data.get("tailoredSkillsParsed") or []:
        skills.append(group.get("name", ""))
        skills.extend(group.get("keywords") or [])
    bullets = []
    for edit in data.get("tailoredExperienceEditsParsed") or []:
        bullets.extend(edit.get("bullets") or [])
    section_order = ((data.get("tailoredLayoutDirectivesParsed") or {}).get("sectionOrder") or [])
    return "\n".join([
        data.get("tailoredHeadline") or "",
        data.get("tailoredSummary") or "",
        *skills,
        *bullets,
        *section_order,
        data.get("tailoredSectionRationale") or "",
        data.get("tailoredOmissionRationale") or "",
    ])


def score_fixture(fixture: dict, data: dict):
    score = 0
    notes = []
    blob = flatten_payload(data)
    expected = fixture["expected"]

    if (data.get("tailoredHeadline") or "").strip() == expected["headline"]:
        score += 2
        notes.append("headline exact")
    elif contains_keyword(data.get("tailoredHeadline") or "", expected["headline"]):
        score += 1
        notes.append("headline partial")
    else:
        notes.append("headline weak")

    required_hits = [kw for kw in expected["requiredKeywords"] if contains_keyword(blob, kw)]
    if len(required_hits) >= max(2, (len(expected["requiredKeywords"]) + 1) // 2):
        score += 2
        notes.append(f"required keywords ok ({len(required_hits)})")
    elif required_hits:
        score += 1
        notes.append(f"required keywords partial ({len(required_hits)})")
    else:
        notes.append("required keywords weak")

    forbidden_hits = [kw for kw in expected["forbiddenKeywords"] if contains_keyword(blob, kw)]
    if not forbidden_hits:
        score += 2
        notes.append("no forbidden claims")
    elif len(forbidden_hits) == 1:
        score += 1
        notes.append(f"one forbidden hit: {forbidden_hits[0]}")
    else:
        notes.append("forbidden hits: " + ", ".join(forbidden_hits))

    rewrite_count = len(data.get("tailoredExperienceEditsParsed") or [])
    if (not expected["shouldRewriteExperience"]) or rewrite_count > 0:
        score += 2
        notes.append(f"experience edits: {rewrite_count}")
    else:
        notes.append("missing experience rewrites")

    actual_order = ((data.get("tailoredLayoutDirectivesParsed") or {}).get("sectionOrder") or [])
    prefix = expected["preferredSectionOrderPrefix"]
    prefix_matches = all(index < len(actual_order) and actual_order[index] == value for index, value in enumerate(prefix))
    if prefix_matches:
        score += 2
        notes.append("section order aligned")
    elif actual_order:
        score += 1
        notes.append("section order partial: " + " > ".join(actual_order))
    else:
        notes.append("missing section order")

    rationale_ok = bool((data.get("tailoredSectionRationale") or "").strip()) and bool((data.get("tailoredOmissionRationale") or "").strip())
    if rationale_ok:
        score += 2
        notes.append("rationales present")
    else:
        notes.append("missing rationale fields")

    return score, notes


def main():
    fixtures = json.loads(FIXTURE_PATH.read_text())
    for fixture in fixtures:
        print("---")
        print(fixture["id"])
        payload = {
            "title": fixture["jobTitle"],
            "employer": "Eval Sample",
            "jobUrl": f"https://example.com/eval/{fixture['id']}",
            "applicationLink": f"https://example.com/eval/{fixture['id']}/apply",
            "location": "Copenhagen",
            "jobDescription": fixture["jobDescription"],
            "status": "discovered",
        }
        create_job = requests.post(f"{BASE_URL}/api/manual-jobs/import", auth=AUTH, json={"job": payload}, timeout=180)
        if not create_job.ok:
            print("FAILED job create", create_job.status_code, create_job.text[:400])
            continue
        job = create_job.json().get("data") or {}
        job_id = job.get("id")
        summarize = requests.post(f"{BASE_URL}/api/jobs/{job_id}/summarize?force=true", auth=AUTH, timeout=180)
        if not summarize.ok:
            print("FAILED summarize", summarize.status_code, summarize.text[:400])
            continue
        data = summarize.json().get("data") or {}
        data["tailoredSkillsParsed"] = json.loads(data.get("tailoredSkills") or "[]") if data.get("tailoredSkills") else []
        data["tailoredExperienceEditsParsed"] = json.loads(data.get("tailoredExperienceEdits") or "[]") if data.get("tailoredExperienceEdits") else []
        data["tailoredLayoutDirectivesParsed"] = json.loads(data.get("tailoredLayoutDirectives") or "{}") if data.get("tailoredLayoutDirectives") else {}
        score, notes = score_fixture(fixture, data)
        print("score=", score, "/ 12")
        print("notes=", " | ".join(notes))
        print("headline=", data.get("tailoredHeadline"))
        print("sectionOrder=", data["tailoredLayoutDirectivesParsed"].get("sectionOrder", []))
        print("experienceEdits=", len(data["tailoredExperienceEditsParsed"]))


if __name__ == "__main__":
    main()

File: scripts/run-tailoring-eval.ts
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { getProfile } from "../orchestrator/src/server/services/profile";
import { generateTailoring } from "../orchestrator/src/server/services/summary";

type Fixture = {
  id: string;
  jobTitle: string;
  jobDescription: string;
  expected: {
    headline: string;
    requiredKeywords: string[];
    forbiddenKeywords: string[];
    shouldRewriteExperience: boolean;
    preferredSectionOrderPrefix: string[];
  };
};

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const fixturePath = path.join(
  __dirname,
  "tailoring-eval-fixtures",
  "denmark-planning.json",
);

function containsKeyword(text: string, keyword: string): boolean {
  return text.toLowerCase().includes(keyword.toLowerCase());
}

function flattenTailoring(
  result: Awaited<ReturnType<typeof generateTailoring>>["data"],
): string {
  if (!result) return "";
  const skills = (result.skills ?? []).flatMap((group) => [
    group.name,
    ...(group.keywords ?? []),
  ]);
  const bullets = (result.experienceEdits ?? []).flatMap(
    (edit) => edit.bullets ?? [],
  );
  const directives = result.layoutDirectives?.sectionOrder ?? [];
  return [
    result.headline,
    result.summary,
    ...skills,
    ...bullets,
    ...directives,
  ].join("\n");
}

function scoreFixture(
  fixture: Fixture,
  data: NonNullable<Awaited<ReturnType<typeof generateTailoring>>["data"]>,
) {
  let score = 0;
  const notes: string[] = [];
  const blob = flattenTailoring(data);

  if (data.headline.trim() === fixture.expected.headline) {
    score += 2;
    notes.push("headline exact");
  } else if (containsKeyword(data.headline, fixture.expected.headline)) {
    score += 1;
    notes.push("headline partial");
  } else {
    notes.push("headline weak");
  }

  const requiredHits = fixture.expected.requiredKeywords.filter((kw) =>
    containsKeyword(blob, kw),
  );
  if (
    requiredHits.length >=
    Math.max(2, Math.ceil(fixture.expected.requiredKeywords.length / 2))
  ) {
    score += 2;
    notes.push(`required keywords ok (${requiredHits.length})`);
  } else if (requiredHits.length > 0) {
    score += 1;
    notes.push(`required keywords partial (${requiredHits.length})`);
  } else {
    notes.push("required keywords weak");
  }

  const forbiddenHits = fixture.expected.forbiddenKeywords.filter((kw) =>
    containsKeyword(blob, kw),
  );
  if (forbiddenHits.length === 0) {
    score += 2;
    notes.push("no forbidden claims");
  } else if (forbiddenHits.length === 1) {
    score += 1;
    notes.push(`one forbidden hit: ${forbiddenHits[0]}`);
  } else {
    notes.push(`forbidden hits: ${forbiddenHits.join(", ")}`);
  }

  const rewriteCount = data.experienceEdits?.length ?? 0;
  if (!fixture.expected.shouldRewriteExperience || rewriteCount > 0) {
    score += 2;
    notes.push(`experience edits: ${rewriteCount}`);
  } else {
    notes.push("missing experience rewrites");
  }

  const actualOrder = data.layoutDirectives?.sectionOrder ?? [];
  const expectedPrefix = fixture.expected.preferredSectionOrderPrefix;
  const prefixMatches = expectedPrefix.every(
    (value, index) => actualOrder[index] === value,
  );
  if (prefixMatches) {
    score += 2;
    notes.push("section order aligned");
  } else if (actualOrder.length > 0) {
    score += 1;
    notes.push(`section order partial: ${actualOrder.join(" > ")}`);
  } else {
    notes.push("missing section order");
  }

  const rationaleOk =
    Boolean(data.sectionRationale?.trim()) &&
    Boolean(data.omissionRationale?.trim());
  if (rationaleOk) {
    score += 2;
    notes.push("rationales present");
  } else {
    notes.push("missing rationale fields");
  }

  return { score, notes };
}

async function main(): Promise<void> {
  const raw = await readFile(fixturePath, "utf8");
  const fixtures = JSON.parse(raw) as Fixture[];
  const profile = await getProfile();

  for (const fixture of fixtures) {
    const result = await generateTailoring(
      {
        jobTitle: fixture.jobTitle,
        jobDescription: fixture.jobDescription,
      },
      profile,
    );
    console.log("---");
    console.log(fixture.id);
    if (!result.success || !result.data) {
      console.log("FAILED", result.error);
      continue;
    }
    const scored = scoreFixture(fixture, result.data);
    console.log("score=", scored.score, "/ 12");
    console.log("notes=", scored.notes.join(" | "));
    console.log("headline=", result.data.headline);
    console.log(
      "sectionOrder=",
      result.data.layoutDirectives?.sectionOrder ?? [],
    );
    console.log("experienceEdits=", result.data.experienceEdits?.length ?? 0);
  }
}

void main();
